File: app/src/main/java/ru/p3tr0vich/fuel/FragmentFuelingRecordChange.java
package ru.p3tr0vich.fuel;

import android.app.Activity;
import android.os.Bundle;
import android.support.annotation.NonNull;
import android.support.annotation.Nullable;
import android.support.v4.app.Fragment;
import android.text.Editable;
import android.text.TextWatcher;
import android.view.LayoutInflater;
import android.view.Menu;
import android.view.MenuInflater;
import android.view.MenuItem;
import android.view.View;
import android.view.ViewGroup;
import android.widget.Button;
import android.widget.EditText;

import com.wdullaer.materialdatetimepicker.date.DatePickerDialog;

import java.util.Calendar;

import ru.p3tr0vich.fuel.helpers.ContentResolverHelper;
import ru.p3tr0vich.fuel.helpers.PreferencesHelper;
import ru.p3tr0vich.fuel.models.FuelingRecord;
import ru.p3tr0vich.fuel.utils.Utils;
import ru.p3tr0vich.fuel.utils.UtilsDate;
import ru.p3tr0vich.fuel.utils.UtilsFormat;

import static ru.p3tr0vich.fuel.R.id.edit_cost;

public class FragmentFuelingRecordChange extends Fragment implements View.OnClickListener {

    public static final String TAG = "FragmentFuelingRecordChange";

    private static final String STATE_KEY_DATE = "STATE_KEY_DATE";
    private static final String STATE_KEY_CALC_STEP = "STATE_KEY_CALC_STEP";

    private FuelingRecord mFuelingRecord;

    private Button mButtonDate;
    private EditText mEditCost;
    private EditText mEditVolume;
    private EditText mEditTotal;

    private float mPrice;

    private PreferencesHelper mPreferencesHelper;

    @NonNull
    public static Fragment newInstance(@Nullable FuelingRecord fuelingRecord) {
        FragmentFuelingRecordChange fragment = new FragmentFuelingRecordChange();

        if (fuelingRecord != null)
            fragment.setArguments(fuelingRecord.toBundle());

        return fragment;
    }

    @Override
    public void onCreate(@Nullable Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        mPreferencesHelper = PreferencesHelper.getInstance(getContext());
    }

    @Override
    public View onCreateView(LayoutInflater inflater, @Nullable ViewGroup container,
                             @Nullable Bundle savedInstanceState) {
        View view = inflater.inflate(R.layout.fragment_fueling_record_change, container, false);

        setHasOptionsMenu(true);

        mButtonDate = (Button) view.findViewById(R.id.btn_date);
        mEditCost = (EditText) view.findViewById(edit_cost);
        mEditVolume = (EditText) view.findViewById(R.id.edit_volume);
        mEditTotal = (EditText) view.findViewById(R.id.edit_total);

        Bundle bundle = getArguments();

        if (bundle != null && bundle.containsKey(FuelingRecord.NAME)) {
            mFuelingRecord = new FuelingRecord(bundle);
            mPrice = 0;
        } else {
            float cost = mPreferencesHelper.getDefaultCost();
            float volume = mPreferencesHelper.getDefaultVolume();
            mPrice = mPreferencesHelper.getPrice();

            if (mPrice != 0) {
                if (cost == 0 && volume != 0) cost = volume * mPrice;
                else if (volume == 0 && cost != 0) volume = cost / mPrice;
            }

            mFuelingRecord = new FuelingRecord(cost, volume, mPreferencesHelper.getLastTotal());
        }

        getActivity().setTitle(mFuelingRecord.getId() != 0 ?
                R.string.dialog_caption_update :
                R.string.dialog_caption_add);

        if (savedInstanceState == null) {
            UtilsFormat.floatToEditText(mEditCost, mFuelingRecord.getCost(), false);
            UtilsFormat.floatToEditText(mEditVolume, mFuelingRecord.getVolume(), false);
            UtilsFormat.floatToEditText(mEditTotal, mFuelingRecord.getTotal(), false);
        } else
            mFuelingRecord.setDateTime(savedInstanceState.getLong(STATE_KEY_DATE));

        updateDate();

        mButtonDate.setOnClickListener(this);

        view.findViewById(R.id.text_cost).setOnClickListener(this);
        view.findViewById(R.id.text_volume).setOnClickListener(this);
        view.findViewById(R.id.text_total).setOnClickListener(this);

        if (savedInstanceState == null)
            mCalcStep = mFuelingRecord.getId() == 0 && mPrice > 0 ? CALC_STEP_SELECT : CALC_STEP_DISABLED;
        else
            mCalcStep = savedInstanceState.getInt(STATE_KEY_CALC_STEP, CALC_STEP_DISABLED);

        return view;
    }

    /**
     * Шаг ожидания ввода.
     * В случае изменения старой записи или если цена меньше или равна нулю
     * вычисление отключено ({@code CALC_STEP_DISABLED}).
     * Иначе, при первом открытии фрагмента, шаг устанавливается на выбор
     * редактируемого и вычисляемого полей ({@code CALC_STEP_SELECT}).
     * При последующих пересозданиях фрагмента загружается предыдущее значение.
     */
    private int mCalcStep;

    /**
     * Выбор редактируемого и вычисляемого полей.
     */
    private static final int CALC_STEP_SELECT = 0;
    /**
     * Редактируемое поле -- Стоимость, вычисляемое поле -- Объём.
     */
    private static final int CALC_STEP_COST = 1;
    /**
     * Редактируемое поле -- Объём, вычисляемое -- Стоимость.
     */
    private static final int CALC_STEP_VOLUME = 2;
    /**
     * Вычисление отключено.
     */
    private static final int CALC_STEP_DISABLED = 3;

    /**
     * Проверка способа изменения поля:
     * true, если поле изменяется в результате ручного ввода,
     * false, если изменение происходит в результате вызова функции setText.
     */
    private boolean mManualInput = true;

    private TextWatcher mCostTextWatcher = null;
    private TextWatcher mVolumeTextWatcher = null;

    @Override
    public void onStart() {
        super.onStart();

        if (mCalcStep != CALC_STEP_DISABLED) {
            if (mCostTextWatcher == null) {
                mCostTextWatcher = new EditTextWatcherAdapter(mEditCost, mEditVolume,
                        new CalcVolume(), CALC_STEP_COST);
                mEditCost.addTextChangedListener(mCostTextWatcher);
            }
            if (mVolumeTextWatcher == null) {
                mVolumeTextWatcher = new EditTextWatcherAdapter(mEditVolume, mEditCost,
                        new CalcCost(), CALC_STEP_VOLUME);
                mEditVolume.addTextChangedListener(mVolumeTextWatcher);
            }
        }
    }

    private class EditTextWatcherAdapter extends TextWatcherAdapter {
        private final EditText mEdited;
        private final EditText mCalculated;

        private final Calculator mCalculator;

        private final int mThisStep;

        /**
         * @param edited     редактируемое поле.
         * @param calculated вычисляемое поле.
         * @param calculator метод вычисления.
         * @param thisStep   редактируемое поле ({@code CALC_STEP_COST или CALC_STEP_VOLUME}).
         */
        public EditTextWatcherAdapter(EditText edited, EditText calculated, Calculator calculator, int thisStep) {
            mEdited = edited;
            mCalculated = calculated;
            mCalculator = calculator;
            mThisStep = thisStep;
        }

        @Override
        public void afterTextChanged(Editable s) {
            // Изначально шаг установлен на выбор.
            // После первого изменения в одном из полей, это поле назначается редактируемым,
            // а другое -- вычисляемым.

            if (mCalcStep == CALC_STEP_SELECT) {
                mCalcStep = mThisStep;
            }

            if (mCalcStep == mThisStep) {
                // Если значение изменяется в редактируемом поле,
                // вычисляется значение для вычисляемого поля.

                float enteredValue = UtilsFormat.editTextToFloat(mEdited);

                float calculatedValue = mCalculator.calc(enteredValue, mPrice);

                mManualInput = false;
                UtilsFormat.floatToEditText(mCalculated, calculatedValue, false);
                mManualInput = true;
            } else {
                // Иначе, если было произведено ручное изменение в вычисляемом поле,
                // вычисление отключается.

                if (mManualInput) {
                    mCalcStep = CALC_STEP_DISABLED;
                    mEditCost.removeTextChangedListener(mCostTextWatcher);
                    mEditVolume.removeTextChangedListener(mVolumeTextWatcher);
                }
            }
        }
    }

    /**
     * Метод вычисления.
     * См. {@link CalcCost}, {@link CalcVolume}
     */
    private interface Calculator {
        float calc(float value, float price);
    }

    /**
     * Вычисляет объём на основе стоимости.
     */
    private static class CalcVolume implements Calculator {
        public float calc(float cost, float price) {
            if (cost == 0 || price == 0) return 0;
            return cost / price;
        }
    }

    /**
     * Вычисляет стоимость на основе объёма.
     */
    private static class CalcCost implements Calculator {
        public float calc(float volume, float price) {
            if (volume == 0 || price == 0) return 0;
            return volume * price;
        }
    }

    @Override
    public void onSaveInstanceState(Bundle outState) {
        super.onSaveInstanceState(outState);

        outState.putLong(STATE_KEY_DATE, mFuelingRecord.getDateTime());
        outState.putInt(STATE_KEY_CALC_STEP, mCalcStep);
    }

    private void updateDate() {
        mButtonDate.setText(UtilsFormat.dateToString(mFuelingRecord.getDateTime(), true));
    }

    @Override
    public void onCreateOptionsMenu(Menu menu, MenuInflater inflater) {
        super.onCreateOptionsMenu(menu, inflater);
        inflater.inflate(R.menu.menu_fueling_record, menu);
    }

    /**
     * Выполняет добавление новой или изменение уже существующей записи.
     *
     * @return Истина, если сохранение успешно.
     */
    private boolean saveRecord() {
        Utils.hideKeyboard(getActivity());

        mFuelingRecord.setCost(UtilsFormat.editTextToFloat(mEditCost));
        mFuelingRecord.setVolume(UtilsFormat.editTextToFloat(mEditVolume));
        mFuelingRecord.setTotal(UtilsFormat.editTextToFloat(mEditTotal));

        if (mFuelingRecord.getId() != 0) {
            if (ContentResolverHelper.updateRecord(getContext(), mFuelingRecord)) {
                return true;
            } else {
                Utils.toast(R.string.message_error_update_record);

                return false;
            }
        } else {
            if (ContentResolverHelper.insertRecord(getContext(), mFuelingRecord)) {
                return true;
            } else {
                Utils.toast(R.string.message_error_insert_record);

                return false;
            }
        }
    }

    private boolean onSaveClicked() {
        if (saveRecord()) {
            mPreferencesHelper.putLastTotal(mFuelingRecord.getTotal());

            Activity activity = getActivity();

            activity.setResult(Activity.RESULT_OK);

            activity.finish();

            return true;
        } else {
            return false;
        }
    }

    @Override
    public boolean onOptionsItemSelected(MenuItem item) {
        switch (item.getItemId()) {
            case R.id.action_save:
                return onSaveClicked();
        }

        return super.onOptionsItemSelected(item);
    }

    @Override
    public void onClick(View v) {
        if (v == mButtonDate) {
            final Calendar calendar = UtilsDate.getCalendarInstance(mFuelingRecord.getDateTime());

            DatePickerDialog.newInstance(
                    new DatePickerDialog.OnDateSetListener() {
                        @Override
                        public void onDateSet(DatePickerDialog view, int year, int monthOfYear, int dayOfMonth) {
                            calendar.set(year, monthOfYear, dayOfMonth);

                            mFuelingRecord.setDateTime(calendar.getTimeInMillis());

                            updateDate();
                        }
                    },
                    calendar.get(Calendar.YEAR),
                    calendar.get(Calendar.MONTH),
                    calendar.get(Calendar.DAY_OF_MONTH)
            ).show(getFragmentManager(), null);
        } else {
            final EditText edit;
            switch (v.getId()) {
                case R.id.text_cost:
                    edit = mEditCost;
                    break;
                case R.id.text_volume:
                    edit = mEditVolume;
                    break;
                case R.id.text_total:
                    edit = mEditTotal;
                    break;
                default:
                    return;
            }
            Utils.showKeyboard(edit);
        }
    }
}

File: app/src/main/kotlin/ru/p3tr0vich/fuel/fragments/FragmentFuelingRecordChange.kt
package ru.p3tr0vich.fuel.fragments

import android.app.Activity
import android.os.Bundle
import android.support.v4.app.Fragment
import android.text.Editable
import android.text.TextWatcher
import android.view.*
import android.widget.Button
import android.widget.EditText
import com.wdullaer.materialdatetimepicker.date.DatePickerDialog
import ru.p3tr0vich.fuel.R
import ru.p3tr0vich.fuel.R.id.edit_cost
import ru.p3tr0vich.fuel.adapters.TextWatcherAdapter
import ru.p3tr0vich.fuel.helpers.ContentResolverHelper
import ru.p3tr0vich.fuel.helpers.PreferencesHelper
import ru.p3tr0vich.fuel.models.FuelingRecord
import ru.p3tr0vich.fuel.utils.Utils
import ru.p3tr0vich.fuel.utils.UtilsDate
import ru.p3tr0vich.fuel.utils.UtilsFormat
import java.util.*

class FragmentFuelingRecordChange : Fragment(), View.OnClickListener {

    private var fuelingRecord: FuelingRecord? = null

    private var buttonDate: Button? = null
    private var editCost: EditText? = null
    private var editVolume: EditText? = null
    private var editTotal: EditText? = null

    private var price = 0f

    private lateinit var preferencesHelper: PreferencesHelper

    /**
     * Шаг ожидания ввода.
     * В случае изменения старой записи или если цена меньше или равна нулю
     * вычисление отключено (`CALC_STEP_DISABLED`).
     * Иначе, при первом открытии фрагмента, шаг устанавливается на выбор
     * редактируемого и вычисляемого полей (`CALC_STEP_SELECT`).
     * При последующих пересозданиях фрагмента загружается предыдущее значение.
     */
    private var calcStep: Int = 0

    /**
     * Проверка способа изменения поля:
     * true, если поле изменяется в результате ручного ввода,
     * false, если изменение происходит в результате вызова функции setText.
     */
    private var manualInput = true

    private var costTextWatcher: TextWatcher? = null
    private var volumeTextWatcher: TextWatcher? = null

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)

        preferencesHelper = PreferencesHelper.getInstance(context!!)
    }

    override fun onCreateView(inflater: LayoutInflater, container: ViewGroup?,
                              savedInstanceState: Bundle?): View? {
        val view = inflater.inflate(R.layout.fragment_fueling_record_change, container, false)

        setHasOptionsMenu(true)

        buttonDate = view.findViewById(R.id.btn_date)
        editCost = view.findViewById(edit_cost)
        editVolume = view.findViewById(R.id.edit_volume)
        editTotal = view.findViewById(R.id.edit_total)

        if (arguments != null && arguments!!.containsKey(FuelingRecord.NAME)) {
            fuelingRecord = FuelingRecord(arguments!!)
            price = 0f
        } else {
            var cost = preferencesHelper.defaultCost
            var volume = preferencesHelper.defaultVolume
            price = preferencesHelper.price

            if (price != 0f) {
                if (cost == 0f && volume != 0f)
                    cost = volume * price
                else if (volume == 0f && cost != 0f) volume = cost / price
            }

            fuelingRecord = FuelingRecord(cost, volume, preferencesHelper.lastTotal)
        }

        activity!!.setTitle(if (fuelingRecord!!.id != 0L)
            R.string.dialog_caption_update
        else
            R.string.dialog_caption_add)

        if (savedInstanceState == null) {
            UtilsFormat.floatToEditText(editCost, fuelingRecord!!.cost, false)
            UtilsFormat.floatToEditText(editVolume, fuelingRecord!!.volume, false)
            UtilsFormat.floatToEditText(editTotal, fuelingRecord!!.total, false)
        } else {
            fuelingRecord!!.dateTime = savedInstanceState.getLong(STATE_KEY_DATE)
        }

        updateDate()

        buttonDate!!.setOnClickListener(this)

        view.findViewById<View>(R.id.text_cost).setOnClickListener(this)
        view.findViewById<View>(R.id.text_volume).setOnClickListener(this)
        view.findViewById<View>(R.id.text_total).setOnClickListener(this)

        calcStep = savedInstanceState?.getInt(STATE_KEY_CALC_STEP, CALC_STEP_DISABLED) ?:
                if (fuelingRecord!!.id == 0L && price > 0) CALC_STEP_SELECT else CALC_STEP_DISABLED

        return view
    }

    override fun onStart() {
        super.onStart()

        if (calcStep != CALC_STEP_DISABLED) {
            if (costTextWatcher == null) {
                costTextWatcher = EditTextWatcherAdapter(editCost, editVolume,
                        CalcVolume(), CALC_STEP_COST)
                editCost!!.addTextChangedListener(costTextWatcher)
            }

            if (volumeTextWatcher == null) {
                volumeTextWatcher = EditTextWatcherAdapter(editVolume, editCost,
                        CalcCost(), CALC_STEP_VOLUME)
                editVolume!!.addTextChangedListener(volumeTextWatcher)
            }
        }
    }

    /**
     * @param edited     редактируемое поле.
     * @param calculated вычисляемое поле.
     * @param calculator метод вычисления.
     * @param thisStep   редактируемое поле (`CALC_STEP_COST или CALC_STEP_VOLUME`).
     */
    private inner class EditTextWatcherAdapter(
            private val edited: EditText?,
            private val calculated: EditText?,
            private val calculator: Calculator,
            private val thisStep: Int) : TextWatcherAdapter() {

        override fun afterTextChanged(s: Editable) {
            // Изначально шаг установлен на выбор.
            // После первого изменения в одном из полей, это поле назначается редактируемым,
            // а другое -- вычисляемым.

            if (calcStep == CALC_STEP_SELECT) {
                calcStep = thisStep
            }

            if (calcStep == thisStep) {
                // Если значение изменяется в редактируемом поле,
                // вычисляется значение для вычисляемого поля.

                val enteredValue = UtilsFormat.editTextToFloat(edited)

                val calculatedValue = calculator.calc(enteredValue, price)

                manualInput = false
                UtilsFormat.floatToEditText(calculated, calculatedValue, false)
                manualInput = true
            } else {
                // Иначе, если было произведено ручное изменение в вычисляемом поле,
                // вычисление отключается.

                if (manualInput) {
                    calcStep = CALC_STEP_DISABLED

                    editCost?.removeTextChangedListener(costTextWatcher)
                    editVolume?.removeTextChangedListener(volumeTextWatcher)
                }
            }
        }
    }

    /**
     * Метод вычисления.
     * См. [CalcCost], [CalcVolume]
     */
    private interface Calculator {
        fun calc(value: Float, price: Float): Float
    }

    /**
     * Вычисляет объём на основе стоимости.
     */
    private class CalcVolume : Calculator {
        override fun calc(value: Float, price: Float): Float {
            return if (value == 0f || price == 0f) 0f else value / price
        }
    }

    /**
     * Вычисляет стоимость на основе объёма.
     */
    private class CalcCost : Calculator {
        override fun calc(value: Float, price: Float): Float {
            return if (value == 0f || price == 0f) 0f else value * price
        }
    }

    override fun onSaveInstanceState(outState: Bundle) {
        super.onSaveInstanceState(outState)

        outState.putLong(STATE_KEY_DATE, fuelingRecord!!.dateTime)
        outState.putInt(STATE_KEY_CALC_STEP, calcStep)
    }

    private fun updateDate() {
        buttonDate!!.text = UtilsFormat.dateToString(fuelingRecord!!.dateTime, true)
    }

    override fun onCreateOptionsMenu(menu: Menu?, inflater: MenuInflater?) {
        super.onCreateOptionsMenu(menu, inflater)
        inflater!!.inflate(R.menu.menu_fueling_record, menu)
    }

    /**
     * Выполняет добавление новой или изменение уже существующей записи.
     *
     * @return Истина, если сохранение успешно.
     */
    private fun saveRecord(): Boolean {
        Utils.hideKeyboard(activity)

        fuelingRecord!!.cost = UtilsFormat.editTextToFloat(editCost)
        fuelingRecord!!.volume = UtilsFormat.editTextToFloat(editVolume)
        fuelingRecord!!.total = UtilsFormat.editTextToFloat(editTotal)

        if (fuelingRecord!!.id != 0L) {
            return if (ContentResolverHelper.updateRecord(context!!, fuelingRecord!!)) {
                true
            } else {
                Utils.toast(R.string.message_error_update_record)

                false
            }
        } else {
            return if (ContentResolverHelper.insertRecord(context!!, fuelingRecord)) {
                true
            } else {
                Utils.toast(R.string.message_error_insert_record)

                false
            }
        }
    }

    private fun onSaveClicked(): Boolean {
        return if (saveRecord()) {
            preferencesHelper.putLastTotal(fuelingRecord!!.total)

            activity!!.setResult(Activity.RESULT_OK)

            activity!!.finish()

            true
        } else {
            false
        }
    }

    override fun onOptionsItemSelected(item: MenuItem?): Boolean {
        when (item!!.itemId) {
            R.id.action_save -> return onSaveClicked()
        }

        return super.onOptionsItemSelected(item)
    }

    override fun onClick(v: View) {
        if (v === buttonDate) {
            val calendar = UtilsDate.getCalendarInstance(fuelingRecord!!.dateTime)

            DatePickerDialog.newInstance(
                    { _, year, monthOfYear, dayOfMonth ->
                        calendar.set(year, monthOfYear, dayOfMonth)

                        fuelingRecord!!.dateTime = calendar.timeInMillis

                        updateDate()
                    },
                    calendar.get(Calendar.YEAR),
                    calendar.get(Calendar.MONTH),
                    calendar.get(Calendar.DAY_OF_MONTH)
            ).show(fragmentManager, null)
        } else {
            val edit = when (v.id) {
                R.id.text_cost -> editCost
                R.id.text_volume -> editVolume
                R.id.text_total -> editTotal
                else -> return
            }

            Utils.showKeyboard(edit)
        }
    }

    companion object {

        const val TAG = "FragmentFuelingRecordChange"

        private const val STATE_KEY_DATE = "STATE_KEY_DATE"
        private const val STATE_KEY_CALC_STEP = "STATE_KEY_CALC_STEP"

        fun newInstance(fuelingRecord: FuelingRecord?): Fragment {
            val fragment = FragmentFuelingRecordChange()

            fragment.arguments = fuelingRecord?.toBundle()

            return fragment
        }

        /**
         * Выбор редактируемого и вычисляемого полей.
         */
        private const val CALC_STEP_SELECT = 0
        /**
         * Редактируемое поле -- Стоимость, вычисляемое поле -- Объём.
         */
        private const val CALC_STEP_COST = 1
        /**
         * Редактируемое поле -- Объём, вычисляемое -- Стоимость.
         */
        private const val CALC_STEP_VOLUME = 2
        /**
         * Вычисление отключено.
         */
        private const val CALC_STEP_DISABLED = 3
    }
}